appengine/genetics/js/blocks.js

```diff
@@ -0,0 +1,464 @@
+/**
+ * Blockly Games: Genetics Blocks
+ *
+ * Copyright 2016 Google Inc.
+ * https://github.com/google/blockly-games
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Blocks for Blockly's Genetics application.
+ * @author kozbial@google.com (Monica Kozbial)
+ */
+'use strict';
+
+goog.provide('Genetics.Blocks');
+
+goog.require('Blockly');
+goog.require('Blockly.JavaScript');
+goog.require('BlocklyGames');
+goog.require('BlocklyGames.JSBlocks');
+
+/**
+ * HSV hue for all genetics mouse function blocks.
+ */
+Genetics.Blocks.GENETICS_MOUSEFUNCTIONS_HUE = 360;
+/**
+ * HSV hue for all genetics blocks (that are not mouse functions).
+ */
+Genetics.Blocks.GENETICS_HUE = 20;
+
+// Extensions to Blockly's language and JavaScript generator.
+
+/**
+ * Generates the JavaScript from a "mouse function" block (pickFight,
+ * chooseMate, mateAnswer) given the name and arguments of the function.
+ * @param {string} funcName The name of the mouse function.
+ * @param {string} args A comma separated string of the argument variable names.
+ * @param {!Blockly.Block} block The block.
+ * @return {Function}
+ * @private
+ */
+Blockly.JavaScript['genetics_generateMouseFunctionJS_'] = function(funcName,
+    args, block) {
+  // Define a procedure with a return value.
+  var branch = Blockly.JavaScript.statementToCode(block, 'STACK');
+  if (Blockly.JavaScript.STATEMENT_PREFIX) {
+    branch = Blockly.JavaScript.prefixLines(
+        Blockly.JavaScript.STATEMENT_PREFIX.replace(/%1/g,
+            '\'' + block.id + '\''), Blockly.JavaScript.INDENT) + branch;
+  }
+  if (Blockly.JavaScript.INFINITE_LOOP_TRAP) {
+    branch = Blockly.JavaScript.INFINITE_LOOP_TRAP.replace(/%1/g,
+        '\'' + block.id + '\'') + branch;
+  }
+  var returnValue = Blockly.JavaScript.valueToCode(block, 'RETURN',
+      Blockly.JavaScript.ORDER_NONE) || '';
+  if (returnValue) {
+    returnValue = '  return ' + returnValue + ';\n';
+  }
+  var code = 'function ' + funcName + '(' + args + ') {\n' +
+      branch + returnValue + '}';
+  code = Blockly.JavaScript.scrub_(block, code);
+  Blockly.JavaScript.definitions_[funcName] = code;
+  return null;
```

Comment:
This is based on the procedure defreturn block. I'm not sure if it is necessary.
@NeilFraser Is it necessary?